File: benchmark/summarize.js
"use strict"

const _             = require('underscore')
const benchmark     = require('./benchmark')
const data          = require('data/data')

const args = benchmark.parseCommandLine()

function summarize(arr) {
  let min
  let max
  const len = arr.length
  let numFinite = 0
  let numNaN = 0
  let sum = 0
  let sum2 = 0

  for (let i = 0; i < len; ++i) {
    const val = arr[i]
    if (Number.isFinite(val)) {
      if (numFinite === 0) min = max = val
      else if (val < min) min = val
      else if (val > max) max = val
      numFinite = numFinite + 1
      sum = sum + val
      sum2 = sum2 + val * val
    }
    else if (isNaN(val)) numNaN = numNaN + 1
  }

  return {len, numFinite, numNaN, min, max, sum, sum2}
}

if (false) {
  // Check result!
  const len = 1024
  const arr = benchmark.randomF64(len)
  console.log(summarize(arr))
}

args.output(args.lengths.map(len => {
  const arr = benchmark.randomF64(len)
  const res = benchmark.time(summarize, arr)
  Object.assign(res, 
    {operation: 'summary', type: 'f64', length: len, mem_size: len * 8})
  return res
}))

File: benchmark/summarize.py
import numpy as np

from   benchmark import time, parse_cmd_line

def np_summarize_0(arr):
    fin = arr[np.isfinite(arr)]
    return dict(
        length      =len(arr),
        numNan      =np.isnan(arr).sum(),
        numFinite   =len(fin),
        min         =fin.min(),
        max         =fin.max(),
        sum         =fin.sum(),
        sum2        =(fin * fin).sum(),
    )

def do(length):
    arr0 = np.random.random(length)
    result = time(np_summarize_0, arr0)
    result.update(
        operation   ="summarize",
        name        ="np_summarize_0", 
        type        ="f64", 
        length      =length, 
        mem_size    =length * 8
    )
    return result


def main():
    args = parse_cmd_line()
    args.output( do(l) for l in args.lengths )


if __name__ == "__main__":
    main()


